File: crates/lure-core/src/provider/openai.rs
//! OpenAI-compatible provider。
//!
//! 对齐上游 `nanobot/providers/openai_compat_provider.py` 的核心请求/响应形状：
//! - 请求体：`{model, messages, temperature, max_tokens}`，POST 到 `{base}/chat/completions`。
//! - 响应：`choices[0].message.content` + `.tool_calls` + `choices[0].finish_reason` + `usage`。
//! - 错误按 HTTP 状态分类为结构化 [`ProviderError`]。
//!
//! `complete_streaming` 走 SSE：请求带 `stream: true`，逐行解析 `data:` 增量并回调，
//! 组装出完整响应（内容/推理拼接、tool_calls 按 index 累积）。
//!
//! Phase 4 不做：`max_completion_tokens`/模型专属覆盖、prompt caching、重试策略
//! （见 upstream-test-ledger）。tool_calls 已解析，供 agent tool-call 循环消费。

use serde_json::{json, Map, Value};

use crate::provider::http::{HttpRequest, HttpResponse, HttpTransport};
use crate::provider::stream::{parse_sse_line, StreamAssembler};
use crate::provider::types::{
    CompletionRequest, GenerationSettings, LlmProvider, LlmResponse, ProviderError, StreamChunk,
    ToolCall,
};
use crate::provider::usage::normalize_usage;

/// 基于 [`HttpTransport`] 的 OpenAI-compatible provider。
pub struct OpenAiCompatProvider<T: HttpTransport> {
    base_url: String,
    api_key: Option<String>,
    model: String,
    provider_name: Option<String>,
    transport: T,
}

impl<T: HttpTransport> OpenAiCompatProvider<T> {
    /// 绑定 base URL、可选 api key、默认模型与传输。
    pub fn new(
        base_url: impl Into<String>,
        api_key: Option<String>,
        model: impl Into<String>,
        transport: T,
    ) -> Self {
        Self {
            base_url: base_url.into(),
            api_key,
            model: model.into(),
            provider_name: None,
            transport,
        }
    }

    /// 绑定 registry provider 名，用于应用 provider 专属请求参数。
    pub fn with_provider_name(mut self, provider_name: impl Into<String>) -> Self {
        self.provider_name = Some(provider_name.into());
        self
    }

    /// 构建 `/chat/completions` 请求（`stream` 控制是否要求 SSE）。
    fn http_request(&self, request: &CompletionRequest, stream: bool) -> HttpRequest {
        let mut body = build_chat_request(&request.model, &request.messages, &request.settings);
        apply_provider_reasoning(&mut body, self.provider_name.as_deref(), &request.settings);
        if !request.tools.is_empty() {
            body["tools"] = Value::Array(request.tools.clone());
        }
        if stream {
            body["stream"] = Value::Bool(true);
            // 请求末帧回传 usage（对齐上游 openai_compat_provider）。
            body["stream_options"] = serde_json::json!({"include_usage": true});
        }
        let url = format!("{}/chat/completions", self.base_url.trim_end_matches('/'));
        let mut headers = vec![("content-type".to_string(), "application/json".to_string())];
        if let Some(key) = &self.api_key {
            headers.push(("authorization".to_string(), format!("Bearer {key}")));
        }
        HttpRequest { url, headers, body }
    }
}

#[async_trait::async_trait]
impl<T: HttpTransport> LlmProvider for OpenAiCompatProvider<T> {
    fn default_model(&self) -> &str {
        &self.model
    }

    async fn complete(&self, request: &CompletionRequest) -> Result<LlmResponse, ProviderError> {
        let http_request = self.http_request(request, false);
        let response = self
            .transport
            .post_json(&http_request)
            .await
            .map_err(ProviderError::Transport)?;

        parse_chat_response(&response)
    }

    async fn complete_streaming(
        &self,
        request: &CompletionRequest,
        on_delta: &mut (dyn FnMut(StreamChunk) + Send),
    ) -> Result<LlmResponse, ProviderError> {
        let http_request = self.http_request(request, true);

        let mut assembler = StreamAssembler::new();
        let mut raw = String::new();
        // 单行解析失败按容错忽略（SSE 常含 keep-alive/注释）；累积原始文本供错误分类。
        let status = self
            .transport
            .post_json_streaming(&http_request, &mut |line| {
                raw.push_str(&line);
                raw.push('\n');
                if let Ok(Some(chunk)) = parse_sse_line(&line) {
                    assembler.push(&chunk);
                    on_delta(chunk);
                }
            })
            .await
            .map_err(ProviderError::Transport)?;

        if let Some(error) = status_error(status, &raw) {
            return Err(error);
        }
        Ok(assembler.finish())
    }
}

/// 非 2xx 状态映射为结构化错误；2xx 返回 `None`。
fn status_error(status: u16, body: &str) -> Option<ProviderError> {
    match status {
        200..=299 => None,
        401 | 403 => Some(ProviderError::Auth {
            status,
            message: snippet(body),
        }),
        429 => Some(ProviderError::RateLimited {
            status,
            message: snippet(body),
        }),
        500..=599 => Some(ProviderError::Server {
            status,
            message: snippet(body),
        }),
        _ => Some(ProviderError::Api {
            status,
            message: snippet(body),
        }),
    }
}

/// 构建 OpenAI-compatible chat completions 请求体。
pub fn build_chat_request(model: &str, messages: &[Value], settings: &GenerationSettings) -> Value {
    json!({
        "model": model,
        "messages": messages,
        "temperature": settings.temperature,
        "max_tokens": settings.max_tokens.max(1),
    })
}

/// 把通用 reasoning 设置映射为 provider 原生的请求形状。
///
/// DeepSeek V4 默认开启思考；`reasoning_effort = "none"` 必须显式发送
/// `thinking.type = "disabled"` 才能切换到非思考模式。关闭时不发送
/// `reasoning_effort`，因为 DeepSeek 只在思考模式接受该力度参数。
fn apply_provider_reasoning(
    body: &mut Value,
    provider_name: Option<&str>,
    settings: &GenerationSettings,
) {
    if !provider_name.is_some_and(|name| name.eq_ignore_ascii_case("deepseek")) {
        return;
    }
    if settings
        .reasoning_effort
        .as_deref()
        .is_some_and(|effort| effort.eq_ignore_ascii_case("none"))
    {
        body["thinking"] = json!({"type": "disabled"});
    }
}

/// 按状态码分类并解析 chat completions 响应。
pub fn parse_chat_response(response: &HttpResponse) -> Result<LlmResponse, ProviderError> {
    let status = response.status;
    let body = &response.body;
    match status {
        200..=299 => {}
        401 | 403 => {
            return Err(ProviderError::Auth {
                status,
                message: snippet(body),
            })
        }
        429 => {
            return Err(ProviderError::RateLimited {
                status,
                message: snippet(body),
            })
        }
        500..=599 => {
            return Err(ProviderError::Server {
                status,
                message: snippet(body),
            })
        }
        _ => {
            return Err(ProviderError::Api {
                status,
                message: snippet(body),
            })
        }
    }

    let value: Value = serde_json::from_str(body)
        .map_err(|e| ProviderError::Response(format!("响应不是合法 JSON: {e}")))?;

    let choice = value
        .get("choices")
        .and_then(Value::as_array)
        .and_then(|choices| choices.first())
        .ok_or_else(|| ProviderError::Response("响应缺少 choices".to_string()))?;

    let message = choice.get("message");
    let content = message
        .and_then(|m| m.get("content"))
        .and_then(Value::as_str)
        .map(str::to_string);
    let reasoning_content = message
        .and_then(|m| m.get("reasoning_content"))
        .and_then(Value::as_str)
        .map(str::to_string);

    let tool_calls = parse_tool_calls(message);

    let finish_reason = choice
        .get("finish_reason")
        .and_then(Value::as_str)
        .unwrap_or("stop")
        .to_string();

    let raw_usage = value
        .get("usage")
        .and_then(Value::as_object)
        .cloned()
        .unwrap_or_else(Map::new);
    let usage = normalize_usage(&raw_usage);

    Ok(LlmResponse {
        content,
        reasoning_content,
        finish_reason,
        usage,
        tool_calls,
    })
}

/// 从 `message.tool_calls` 解析工具调用；缺失或非数组时返回空。
fn parse_tool_calls(message: Option<&Value>) -> Vec<ToolCall> {
    message
        .and_then(|m| m.get("tool_calls"))
        .and_then(Value::as_array)
        .map(|calls| {
            calls
                .iter()
                .filter_map(|call| {
                    let function = call.get("function")?;
                    let name = function.get("name").and_then(Value::as_str)?;
                    let id = call.get("id").and_then(Value::as_str).unwrap_or_default();
                    let arguments = function
                        .get("arguments")
                        .and_then(Value::as_str)
                        .unwrap_or_default();
                    Some(ToolCall {
                        id: id.to_string(),
                        name: name.to_string(),
                        arguments: arguments.to_string(),
                    })
                })
                .collect()
        })
        .unwrap_or_default()
}

/// 截断过长的错误响应体，避免污染错误消息。
fn snippet(body: &str) -> String {
    const MAX: usize = 500;
    if body.chars().count() <= MAX {
        return body.to_string();
    }
    let head: String = body.chars().take(MAX).collect();
    format!("{head}…")
}

File: packages/core/src/provider/openai.ts
//! OpenAI-compatible provider（对齐 crates/lure-core/src/provider/openai.rs）。

import type {
  CompletionRequest,
  GenerationSettings,
  Json,
  LlmResponse,
  StreamChunk,
  ToolCall,
} from "./types.js";
import { ProviderError } from "./types.js";
import type { HttpRequest, HttpResponse, HttpTransport } from "./http.js";
import { asArray, asObject, asString } from "./json.js";
import { parseSseLine, StreamAssembler } from "./stream.js";
import { normalizeUsage } from "./usage.js";

export class OpenAiCompatProvider {
  constructor(
    private readonly baseUrl: string,
    private readonly apiKey: string | undefined,
    private readonly model: string,
    private readonly transport: HttpTransport,
    private providerName?: string,
  ) {}

  withProviderName(name: string): this {
    this.providerName = name;
    return this;
  }

  defaultModel(): string {
    return this.model;
  }

  private httpRequest(request: CompletionRequest, stream: boolean): HttpRequest {
    const body = buildChatRequest(request.model, request.messages, request.settings);
    applyProviderReasoning(body, this.providerName, request.settings);
    if (request.tools.length > 0) {
      body["tools"] = request.tools;
    }
    if (stream) {
      body["stream"] = true;
      body["stream_options"] = { include_usage: true };
    }
    const url = `${this.baseUrl.replace(/\/+$/, "")}/chat/completions`;
    const headers: [string, string][] = [["content-type", "application/json"]];
    if (this.apiKey !== undefined) {
      headers.push(["authorization", `Bearer ${this.apiKey}`]);
    }
    return { url, headers, body };
  }

  async complete(request: CompletionRequest): Promise<LlmResponse> {
    const httpRequest = this.httpRequest(request, false);
    let response: HttpResponse;
    try {
      response = await this.transport.postJson(httpRequest);
    } catch (e) {
      throw new ProviderError("transport", String(e));
    }
    const llm = parseChatResponse(response);
    logLlmResponse(llm);
    return llm;
  }

  async completeStreaming(
    request: CompletionRequest,
    onDelta: (chunk: StreamChunk) => void,
  ): Promise<LlmResponse> {
    const httpRequest = this.httpRequest(request, true);

    const assembler = new StreamAssembler();
    let raw = "";
    let status: number;
    try {
      status = await this.transport.postJsonStreaming(httpRequest, (line) => {
        raw += line + "\n";
        try {
          const chunk = parseSseLine(line);
          if (chunk !== null) {
            assembler.push(chunk);
            onDelta(chunk);
          }
        } catch {
          // 单行解析失败按容错忽略（SSE 常含 keep-alive/注释）。
        }
      });
    } catch (e) {
      throw new ProviderError("transport", String(e));
    }

    const error = statusError(status, raw);
    if (error) throw error;

    const llm = assembler.finish();
    logLlmResponse(llm);
    return llm;
  }
}

/// 构建 OpenAI-compatible chat completions 请求体。
export function buildChatRequest(
  model: string,
  messages: Json[],
  settings: GenerationSettings,
): Record<string, Json> {
  return {
    model,
    messages,
    temperature: settings.temperature,
    max_tokens: Math.max(1, settings.maxTokens),
  };
}

/// 把通用 reasoning 设置映射为 provider 原生的请求形状（DeepSeek thinking 开关）。
function applyProviderReasoning(
  body: Record<string, Json>,
  providerName: string | undefined,
  settings: GenerationSettings,
): void {
  if (providerName === undefined || providerName.toLowerCase() !== "deepseek") return;
  if (settings.reasoningEffort?.toLowerCase() === "none") {
    body["thinking"] = { type: "disabled" };
  }
}

/// 按状态码分类并解析 chat completions 响应。
export function parseChatResponse(response: HttpResponse): LlmResponse {
  const error = statusError(response.status, response.body);
  if (error) throw error;

  let value: Json;
  try {
    value = JSON.parse(response.body);
  } catch (e) {
    throw new ProviderError("response", `响应不是合法 JSON: ${e}`);
  }

  const root = asObject(value) ?? {};
  const choice = asArray(root["choices"])?.[0];
  if (choice === undefined) {
    throw new ProviderError("response", "响应缺少 choices");
  }

  const message = asObject(asObject(choice)?.["message"]);
  const content = asString(message?.["content"]);
  const reasoningContent = asString(message?.["reasoning_content"]);
  const toolCalls = parseToolCalls(message);
  const finishReason = asString(asObject(choice)?.["finish_reason"]) ?? "stop";
  const usage = normalizeUsage(asObject(root["usage"]) ?? {});

  return { content, reasoningContent, finishReason, usage, toolCalls };
}

/// 非 2xx 状态映射为结构化错误；2xx 返回 `undefined`。
function statusError(status: number, body: string): ProviderError | undefined {
  if (status >= 200 && status <= 299) return undefined;
  if (status === 401 || status === 403) return new ProviderError("auth", snippet(body), status);
  if (status === 429) return new ProviderError("rate_limited", snippet(body), status);
  if (status >= 500 && status <= 599) return new ProviderError("server", snippet(body), status);
  return new ProviderError("api", snippet(body), status);
}

/// 从 `message.tool_calls` 解析工具调用；缺失或非数组时返回空。
function parseToolCalls(message: Record<string, Json> | undefined): ToolCall[] {
  const calls = asArray(message?.["tool_calls"]);
  if (!calls) return [];
  return calls.flatMap((call) => {
    const callObj = asObject(call);
    const fn = asObject(callObj?.["function"]);
    const name = asString(fn?.["name"]);
    if (name === undefined) return [];
    const id = asString(callObj?.["id"]) ?? "";
    const args = asString(fn?.["arguments"]) ?? "";
    return [{ id, name, arguments: args }];
  });
}

/// 截断过长的错误响应体，避免污染错误消息。
function snippet(body: string): string {
  const MAX = 500;
  if (body.length <= MAX) return body;
  return `${body.slice(0, MAX)}…`;
}

/// 调试打印 provider 返回的完整结果（think/content/tool_call）。
function logLlmResponse(response: LlmResponse): void {
  if (response.reasoningContent) {
    console.error(`[llm] think:\n${response.reasoningContent}`);
  }
  if (response.content) {
    console.error(`[llm] content:\n${response.content}`);
  }
  response.toolCalls.forEach((tool, i) => {
    console.error(`[llm] tool_call[${i}] id=${tool.id} name=${tool.name}\n  args: ${tool.arguments}`);
  });
  console.error(`[llm] finish_reason=${response.finishReason}`);
}
